main_boaz.tex

```diff
@@ -0,0 +1,910 @@
+
+
+We also allow \emph{partial axising}:
+\begin{align*}
+A_{\nidx{foo}{1}} &= \nmatrix{}{bar}{
+  3 & 1 & 4
+}
+\\
+A_{\nidx{bar}{3}} &= \nmatrix{}{foo}{
+  4 & 9
+}.
+\end{align*}
+
+In general, if $A$ is a tensor of shape $\mathcal{S}$ where $d=|\mathcal{S}|$, and $i_1,\ldots,i_\ell$ are elements in distinct slices $S_1,\ldots,S_\ell \in \mathcal{S}$ then $A(i_1,\ldots,i_\ell)$ is the tensor of shape $\mathcal{S} \setminus \{ S_1,\ldots, S_\ell \}$ that maps $j_1,\ldots,j_d-{\ell}$ into the $A(i_1,\ldots,i_\ell,j_1,\ldots,j_{d-\ell})$.
+
+
+
+\paragraph{Proposed convention:} We propose that if $i$ is an axis belonging to an axis  \emph{not} in  $type(A)$ then $A(i)=A$.  This can be convenient for example to use a $\{ \name{width} ,\name{height} \}$ type tensor in a function that expects a $\{ \name{width} , \name{height} , \name{batch} \}$ tensor by essentially treating the former as if it has a trivial (only one coordinate) $\name{batch}$ axis.
+
+
+\subsection{Named tensor operations}
+\label{sec:operations}
+
+\subsection{General operations}
+
+An \emph{operator}  takes as input one or more tensors and outputs as input a tensor. For example:
+
+$CONV: \mathbb{R}^{\name{width},\name{height}} \times \mathbb{R}^{\name{width},\name{height}} \rightarrow \mathbb{R}^{\name{width},\name{height}}$
+
+is the two dimensional convolution defined as following: $CONV(W,X)=Y$ of shape $\{ \name{width}(X),\name{height}(X) \}$ such that for every $(i,j) \in (\name{width},\name{height})$, 
+
+$Y_{x,y} = \sum_{(a,b) \in (\name{width}(W),\name{height}(W)} X_{x+a,y+b} W_{a,b}$
+
+Another way to write it is that 
+
+$Y_{x,y} =  X_{\name{width}[x-\ell:x+\ell],\name{height}[y-\ell,y+\ell]} \cdot W$ where $|\name{width}(W)|=|\name{height}(W)|=2\ell+1$.
+
+(We are thinking of the axes of $W$ here as indexed by integers $\{ -\ell, -\ell+1, \ldots, 0, 1,\ldots, \ell \}$. We assume here that any ``out of bound'' value defaults to zero.)
+
+\paragraph{Dangling and aligned axes} We use the following conventions for an operator $F$ taking one or more tensors $X_1,X_2,\ldots,X_\ell$ when these tensors contain axes that do not appear in the signature of $F$:
+
+\begin{itemize}
+  \item If an axis $\name{ax}$ appears only in one of the $X_i$'s and not in any other, then it is called \emph{dangling}. In such a case, we execute $F$ in parallel for every slice of the form $X_i(\name{ax}=i)$ and the output tensor will contain the axis $\name{ax}$ with the same support as that of this $X_i$ with the corresponding slice having the output of $F$ on that slice.
+
+  \item If an axis $\name{ax}$ that does not appear in the signature appears in more than one of the $X_i$'s, then the two appearances must have the same support (unless we explain how aligning them can happen). In such a case these axes are \emph{aligned}. In such cases, for every $i$ in the support of this axis, we run $F$ on the corresponding slices of all tensors in which the axis appears.
```

Comment:
If

A ∈ R^{ax^1}
B ∈ R^{ax^2}

in the expression A+B, are both ax^1 and ax^2 dangling? (I hope yes.)

Reply:
Sorry the way I view it is that there is automatic casting and so if we add these two vectors we will simply add them.

One way to think about this is the following. 
Suppose you have a transformation F on images that takes as { **width**^in , **height**^in }  to { **width**^out , **height**^out }  and a transformation G that takes two images (both thought of as { **width** , **height** } ) and combines them together. You sometimes do these type of things in various data augmentation procedure.

If A is an image, we might want to run G(A,F(A)) and it would treat the axes of A and F(A) as aligned.

I think in situations as the one that you are describing, it's better to express exactly which axes in the + you are planning to align and which not. Maybe you can write an empty set below the + if you wish to align no axes.

Reply:
OK, then these rules need to clarify that automatic casting can occur when checking whether an axis appears in more than one argument. This raises more questions, like:

- ax^1 and ax^2 don't actually have the same support, because ax^1 consists of triples (ax, 1, ...) and ax^2 consists of triples (ax, 2, ...). So the rule needs to say that they have the same support after casting.

- If A ∈ R^{ax^1, ax^2} and B ∈ R^{ax^2}, then presumably A+B should align ax^2 with ax^2? The rules would need to prevent us from looking for a match for A's ax^1 first, incorrectly aligning it with B's ax^2 and then leaving A's ax^2 dangling.

- If A ∈ R^{ax^1, ax^2} and B ∈ R^{ax}, then what is A+B? If I look at A's ax^1 first, I can align it with B's ax. Then if I look at A's ax^2, I'm not sure if it will dangle because B's ax is already "taken," or whether it can align with B's ax also. On the other hand, if I look at B's ax first, then it will fail because A has two potentially alignable axes.

I'm guessing that there are more edge cases to consider...I hope you get my point that I think the automatic casting rules will get very complicated if written out fully (as I believe they need to be).

Reply:
The principle in the notation would be::

1) There is no order in tensor axes

2) If an expression would give different results depending on which order you process indices then it is undefined.

3) You always look for the more specific match first before the less specific one (similar to philosophy of Julia's multiple dispatch). So if you have ax^1 and ax^2 and B is in ax^2 then then yes ax^1 would be dangling. 

4) There is always a way for the author to explicitly state exactly what they want to align, and to state what the type of the result will look like 



Writing rules formally might be cumbersome (just like Julia's multiple dispatch.) but I think that overall the notation in papers should make it clear what's going on. 

No one really needs to think about triples - this was just for formal definition and casting would modify the second element - the point is that we want to distinguish between two different and common occurrences in usage of tensors:

1) When two axes have truly different types, such as batches, channels, time, width, height, etc. 

2) When we want to be able to refer to more than one copy of the same type of axis. Often in this case these distinguishing marks are "ad hoc" and temporary, and the same tensor that plays the role  of (for example) **state**^in in one operation, would play the role of **state**^out in another.

These two are different uses, and this is why the type + annotation  comes from.  Frankly, if all we had to deal was case #2, then we probably would have been fine with the standard un-named tensor convention which just distinguishes between them based on order. 

Here is a worked out example:
![image](https://user-images.githubusercontent.com/5474168/101654469-904f4780-3a0e-11eb-8afd-2d3725da0447.png)